File: JavaScript/Medium/SumOfTwoIntegers.js
/**
 * @param {number} a
 * @param {number} b
 * @return {number}
 */
 var getSum = function(a, b) {
  
    while (b !== 0) {
		let temp = a;
		a = a ^ b;
		b = (temp & b) << 1;
	}
    
	return a;  
};

/*

    Example:
    a = 1010
    b = 0011
    
    1st iteration:
    temp = 1010
    a = 1001
    b = 0100
    
    2nd iteration:
    temp = 1001
    a = 1101
    b = 0000
    
    answer = 1101 = 13
    
    Time Complexity: O(1)
    Space Complexty: O(1)

*/

File: Swift/Medium/SumOfTwoIntegers.swift
class Solution {
    func getSum(_ a: Int, _ b: Int) -> Int {
        if (b == 0) { return a }
        return getSum(a ^ b, (a & b) << 1)
    }
}

/*

    Recursive

    Time: O(1)
    Space: O(1)

    Constant time and space because the Int type is either a 32-bit integer or 64-bit integer

*/
